firedrake/preconditioners/fdm.py

```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
+        degree = V.finat_element.degree
+        try:
+            degree = max(degree)
+        except TypeError:
+            pass
+        if formdegree == tdim:
+            degree = degree + 1
+        is_interior, is_facet = is_restricted(V.finat_element)
+        key = (degree, tdim, formdegree, V.value_size, is_interior, is_facet)
+        cache = self._reference_tensor_cache
+        try:
+            return cache[key]
+        except KeyError:
+            full_key = (degree, tdim, formdegree, V.value_size, False, False)
+            if is_facet and full_key in cache:
+                result = cache[full_key]
+                noperm = PETSc.IS().createGeneral(numpy.arange(result.getSize()[0], dtype=PETSc.IntType), comm=result.comm)
+                result = result.createSubMatrix(noperm, self.ises[1])
+                noperm.destroy()
+                return cache.setdefault(key, result)
+
+            elements = sorted(get_base_elements(V.finat_element), key=lambda e: e.formdegree)
+            ref_el = elements[0].get_reference_element()
+            eq = FIAT.FDMQuadrature(ref_el, degree)
+            e0 = elements[0] if elements[0].formdegree == 0 else FIAT.FDMLagrange(ref_el, degree)
+            e1 = elements[-1] if elements[-1].formdegree == 1 else FIAT.FDMDiscontinuousLagrange(ref_el, degree-1)
+            if is_interior:
+                e0 = FIAT.RestrictedElement(e0, restriction_domain="interior")
+            if hasattr(eq.dual, "rule"):
+                rule = eq.dual.rule
+            else:
+                rule = FIAT.quadrature.make_quadrature(ref_el, degree+1)
+
+            pts = rule.get_points()
+            wts = rule.get_weights()
+
+            phiq = eq.tabulate(0, pts)
+            phi1 = e1.tabulate(0, pts)
+            phi0 = e0.tabulate(1, pts)
+
+            moments = lambda v, u: numpy.dot(numpy.multiply(v, wts), u.T)
+            A00 = moments(phiq[(0, )], phi0[(0, )])
+            A11 = moments(phi1[(0, )], phi1[(0, )])
+            A10 = moments(phi1[(0, )], phi0[(1, )])
+            A10 = numpy.linalg.solve(A11, A10)
+            A11 = numpy.eye(A11.shape[0])
+
+            Ihat = mass_matrix(tdim, formdegree, A00, A11)
+            Dhat = diff_matrix(tdim, formdegree, A00, A11, A10)
+            result = block_mat([[Ihat], [Dhat]])
+            Ihat.destroy()
+            Dhat.destroy()
+
+            if value_size != 1:
+                eye = petsc_sparse(numpy.eye(value_size))
+                temp = result
+                result = temp.kron(eye)
+                temp.destroy()
+                eye.destroy()
+
+            if is_facet:
+                cache[full_key] = result
+                noperm = PETSc.IS().createGeneral(numpy.arange(result.getSize()[0], dtype=PETSc.IntType), comm=result.comm)
+                result = result.createSubMatrix(noperm, self.ises[1])
+                noperm.destroy()
+
+            return cache.setdefault(key, result)
+
+
+def factor_interior_mat(A00):
```

Comment:
Comment "Used in static condensation. Take in A00 on a cell, return its Cholesky factorisation."